misc/userscripts/password_fill

```diff
@@ -0,0 +1,276 @@
+#    are no requirements how these names should look like.
+#  - open_entry() is called with some specific entry of the $files array and is
+#    expected to write the username of that entry to the $username variable and
+#    the corresponding password to $password
+
+reset_backend() {
+    init() { true ; }
+    query_entries() { true ; }
+    open_entry() { true ; }
+}
+
+# choose_entry() is expected to choose one entry from the array $files and
+# write it to the variable $file.
+choose_entry() {
+    choose_entry_random
+}
+
+# The default implementation chooses a random entry from the array. So if there
+# are multiple matching entries, multiple calls to this userscript will
+# eventually pick the "correct" entry. I.e. if this userscript is bound to
+# "zl", the user has to press "zl" until the correct username shows up in the
+# login form.
+choose_entry_random() {
+    local nr=${#files[@]}
+    file="${files[$((RANDOM % nr))]}"
+    # Warn user, that there might be other matching password entries
+    if [ "$nr" -gt 1 ] ; then
+        msg "Picked $file out of $nr entries: ${files[*]}"
+    fi
+}
+
+# another implementation would be to ask the user via some menu (like rofi or
+# dmenu or zenity or even qutebrowser completion in future?) which entry to
+# pick
+MENU_COMMAND=( head -n 1 )
+choose_entry_menu() {
+    local nr=${#files[@]}
+    if [ "$nr" -eq 1 ] ; then
+        file="${files[0]}"
+    else
+        file=$( printf "%s\n" "${files[@]}" | "${MENU_COMMAND[@]}" )
+    fi
+}
+
+choose_entry_rofi() {
+    MENU_COMMAND=( rofi -p "qutebrowser> " -dmenu
+                        -mesg $'Pick a password entry for <b>'"${QUTE_URL//&/&amp;}"'</b>' )
+    choose_entry_menu || true
+}
+
+choose_entry_zenity() {
+    MENU_COMMAND=( zenity --list --title "Qutebrowser password fill"
+                          --text "Pick the password entry:"
+                          --column "Name" )
+    choose_entry_menu || true
+}
+
+choose_entry_zenity_radio() {
+    zenity_helper() {
+        awk '{ print $0 ; print $0 }'                   \
+        | zenity --list --radiolist                     \
+                 --title "Qutebrowser password fill"    \
+                 --text "Pick the password entry:"      \
+                 --column " " --column "Name"
+    }
+    MENU_COMMAND=( zenity_helper )
+    choose_entry_menu || true
+}
+
+# =======================================================
+# backend: PASS
+
+# configuration options:
+match_filename=1 # whether allowing entry match by filepath
+match_line=1     # whether allowing entry match by URL-Pattern in file
+match_line_pattern='^url: .*' # applied using grep -iE
+user_pattern='^(user|username): '
+
+GPG_OPTS=( "--quiet" "--yes" "--compress-algo=none" "--no-encrypt-to" )
+GPG="gpg"
+export GPG_TTY="${GPG_TTY:-$(tty 2>/dev/null)}"
+which gpg2 &>/dev/null && GPG="gpg2"
+[[ -n $GPG_AGENT_INFO || $GPG == "gpg2" ]] && GPG_OPTS+=( "--batch" "--use-agent" )
+
+pass_backend() {
```

Comment:
Right. I directly access passes backend, because the output of the `pass` command is not nicely parsable for scripts (especially that of `pass grep`).